### PR TITLE
Signup: Check all Signup flows for a valid configuration

### PR DESCRIPTION
**Pull request by bisko**
This PR introduces basic unit testing for the Signup flows. 

This will help reveal errors in the Signup flows configuration that may appear from time to time and can be a bit obscure if a specific flow isn't covered by the E2E tests.

The PR doesn't cover all possible flow configuration tests and only verifies that the step dependencies, specified in the config, are met. 

There are other tests that can be implemented, to properly test each flow from start to finish, to verify they are working correctly, but it requires improving the Signup framework to allow for proper unit-testing and is out of scope for the moment. 

Since the PR introduces some reorganization in the configuration files it would require testing.

To test:

1. Checkout branch or use Calypso.live link
2. Go through several flows in Signup, e.g. `main`, `premium`, the Jetpack flows, `get.blog` flows
3. Verify Signup is working correctly
4. Verify no errors in JS console

### CODE REVIEW

File: client/signup/main.jsx [hunk 1]
```diff
@@ -139,25 +139,35 @@ const Signup = React.createClass( {
...
+							this.handleFlowComplete.bind( this, dependencies, filteredDestination ),
+							MINIMUM_TIME_LOADING_SCREEN_IS_DISPLAYED - timeSinceLoading
+						);
+					}
+					return this.handleFlowComplete( dependencies, filteredDestination );
+				}.bind( this )
+			} );
+		} catch ( Exception ) {
+
+			// TODO fixme!
```
Comment (lsinger) [inline client/signup/main.jsx]: Can you describe briefly what's to fix (why, how)?

File: client/signup/main.jsx [hunk 2]
```diff
@@ -139,25 +139,35 @@ const Signup = React.createClass( {
...
-				}
-				return this.handleFlowComplete( dependencies, filteredDestination );
-			}.bind( this )
-		} );
+		try {
+			this.signupFlowController = new SignupFlowController( {
+				flowName: this.props.flowName,
+				providedDependencies,
+				reduxStore: this.context.store,
+				onComplete: function( dependencies, destination ) {
```
Comment (lsinger) [inline client/signup/main.jsx]: moving the definition of this function out of this constructor could help make it more readable.

**Comment by lsinger** [comment]
Signing up via `/start` works fine, but starting with e.g. `/start/jetpack-user` (which I thought should work?) gives me an error in the console and a blank Calypso: 

`TypeError: Flows.insertStepIntoFlow is not a function`

File: client/signup/config/test/index.js [hunk 3]
```diff
@@ -32,3 +36,64 @@ describe( 'index', () => {
...
+					DependencyStore.reset();
+					ProgressStore.reset();
+				} );
+
+				it( 'Valid configuration', () => {
+					try {
+						flowControllerInstance = FlowController( {
+							flowName: flowName,
+							onComplete: () => {
+								console.log( 'Complete flow' );
```
Comment (mcsf) [inline client/signup/config/test/index.js]: I was a little thrown off by this line. Removing the `log` statement didn't change the test outcome for me (I think).

What I then did try, due to seeing the name `onComplete`, was the following:

```diff
diff --git a/client/signup/config/test/index.js b/client/signup/config/test/index.js
index ee4bb5ec2..0caa98d37 100644
--- a/client/signup/config/test/index.js
+++ b/client/signup/config/test/index.js
@@ -75,12 +75,12 @@ describe( 'Signup flows configuration validation', () => {
 					ProgressStore.reset();
 				} );
 
-				it( 'Valid configuration', () => {
+				it( 'Valid configuration', ( done ) => {
 					try {
 						flowControllerInstance = FlowController( {
 							flowName: flowName,
 							onComplete: () => {
-								console.log( 'Complete flow' );
+								done();
 							}
 						} );
 					} catch ( Exception ) {
```

but this results in all tests timing out. As I'm mostly unfamiliar with Signup, my questions are:

- Do we not expect the controller instance to _complete_?
- If we don't, do we need to define a `onComplete` function? If so, can it be `noop`?
- If we don't really reach `onComplete`, then we're really mostly testing by making sure no errors are thrown in that same stack. Layman's question: is that enough?

**Comment by bisko** [comment]
Closing this, since it's old and things have changed quite a lot since last working on it. May revisit in the future, if we shit focus back to Signup and customer acquisition.

**Pull request closed without merging**